### ISSUE TITLE
Foo_Bar is not a valid hostname

### ISSUE THREAD

**Issue by garethsb-sony** [sony/nmos-cpp#11]
nmos-cpp-registry rejects node registrations with "Foo_Bar is not a valid hostname", when the value contains an underscore.
 
This is due to this code, which rejects the underscore:

https://github.com/sony/nmos-cpp/blob/e78b0851b5ed689b6b19b463df836edef1f41026/Development/cpprest/json_validator_impl.cpp#L46-L52

That implementation was based on https://stackoverflow.com/a/106223, which quotes RFC1123 and RFC952.

However, the accepted answer to https://stackoverflow.com/questions/2180465/can-domain-name-subdomains-have-an-underscore-in-it quotes RFC2181, which does allow underscores in domain names.

So, we need to understand what "hostname" is actually intended to mean!

Thanks to @billt-hlit for discovering this issue.

**Comment by garethsb-sony** [comment]
Some more references:

- [JSON Schema 7.3.3. Hostnames](http://json-schema.org/latest/json-schema-validation.html#rfc.section.7.3.3), which actually refers to [RFC 1034, Section 3.1](https://tools.ietf.org/html/rfc1034#section-3.1)
- However, actually only [Section 3.5](https://tools.ietf.org/html/rfc1034#section-3.5) has any character constraints, and that is only a "Preferred name syntax" - it doesn't allow underscore
- I found a similar issue reported on another json-schema implementation here: https://github.com/justinrainbow/json-schema/issues/129

**Comment by garethsb-sony** [comment]
@andrewbonney, do you have any  further thoughts on this?

**Comment by garethsb-sony** [comment]
Regarding DNS-SD, it's true that SRV records make use of "_", but isn't this largely to distinguish those names from 'real' hostnames? See [RFC2782](https://tools.ietf.org/html/rfc2782) which contains:

> An underscore (_) is prepended to the service identifier to avoid collisions with DNS labels that occur in nature.

The number of possibly relevant RFCs is bewildering, but it certainly looks like several platforms do allow underscores in hostnames, so I agree it's probably the case that a registry should allow them...

On the other hand, if we only implement the base DNS spec, doesn't that just mean just 1-63 octets (of any value, 0-255, possibly except '.'?) per label, separated by '.', and, I think, a max length of 255 octets in total?

I'm tempted to give up, and do no extra validation...

**Comment by andrewbonney** [comment]
I agree, otherwise we're likely to be chasing different interpretations in different implementations forever.
